broker/src/main/java/io/zeebe/broker/exporter/metrics/MetricsExporter.java

```diff
@@ -0,0 +1,154 @@
+    jobKeyToCreationTimeMap.clear();
+    workflowInstanceKeyToCreationTimeMap.clear();
+  }
+
+  @Override
+  public void export(final Record record) {
+    if (record.getRecordType() != RecordType.EVENT) {
+      controller.updateLastExportedRecordPosition(record.getPosition());
+      return;
+    }
+
+    final var partitionId = record.getPartitionId();
+    final var recordKey = record.getKey();
+
+    final var currentValueType = record.getValueType();
+    if (currentValueType == ValueType.JOB) {
+      handleJobRecord(record, partitionId, recordKey);
+    } else if (currentValueType == ValueType.WORKFLOW_INSTANCE) {
+      handleWorkflowInstanceRecord(record, partitionId, recordKey);
+    }
+
+    controller.updateLastExportedRecordPosition(record.getPosition());
+  }
+
+  private void handleWorkflowInstanceRecord(
+      final Record<?> record, final int partitionId, final long recordKey) {
+    final var currentIntent = record.getIntent();
+
+    if (currentIntent == WorkflowInstanceIntent.ELEMENT_ACTIVATING
+        && isWorkflowInstanceRecord(record)) {
+      storeWorkflowInstanceCreation(record.getTimestamp(), recordKey);
+    } else if (currentIntent == WorkflowInstanceIntent.ELEMENT_COMPLETED
+        && isWorkflowInstanceRecord(record)) {
+      final var creationTime = workflowInstanceKeyToCreationTimeMap.remove(recordKey);
+      executionLatencyMetrics.observeWorkflowInstanceExecutionTime(
+          partitionId, creationTime, record.getTimestamp());
+    }
+  }
+
+  private void storeWorkflowInstanceCreation(final long creationTime, final long recordKey) {
+    workflowInstanceKeyToCreationTimeMap.put(recordKey, creationTime);
+    creationTimeToWorkflowInstanceKeyTreeMap.put(creationTime, recordKey);
+  }
+
+  private void handleJobRecord(
+      final Record<?> record, final int partitionId, final long recordKey) {
+    final var currentIntent = record.getIntent();
+
+    if (currentIntent == JobIntent.CREATED) {
+      storeJobCreation(record.getTimestamp(), recordKey);
+    } else if (currentIntent == JobIntent.ACTIVATED) {
+      final var creationTime = jobKeyToCreationTimeMap.get(recordKey);
+      executionLatencyMetrics.observeJobActivationTime(
+          partitionId, creationTime, record.getTimestamp());
+    } else if (currentIntent == JobIntent.COMPLETED) {
+      final var creationTime = jobKeyToCreationTimeMap.remove(recordKey);
+      executionLatencyMetrics.observeJobLifeTime(partitionId, creationTime, record.getTimestamp());
+    }
+  }
+
+  private void storeJobCreation(final long creationTime, final long recordKey) {
+    jobKeyToCreationTimeMap.put(recordKey, creationTime);
+    creationTimeToJobKeyTreeMap.put(creationTime, recordKey);
+  }
+
+  private void cleanUp() {
+    final var currentTimeMillis = System.currentTimeMillis();
+
+    final var deadTime = currentTimeMillis - TIME_TO_LIVE.toMillis();
+    clearMaps(deadTime, creationTimeToJobKeyTreeMap, jobKeyToCreationTimeMap);
```

Comment:
> If we write a batch of follow up events, it's likely they all have the same creation time, no?

I think they have the same time on a batch.

> Could we end up with "garbage" over time? I put it in quotes since technically we should eventually see the completion record for that entity, but still.

Can you elaborate on that?

Reply:
So I add a bunch of keys with the same creation time into the `creationTimeToJobKeyTreeMap` (or the other one). It's a map, so it only saves the latest value, no? So when I clean up, I actually only clean up that value, not any of the others from the map? Or am I missing something?

Reply:
Ok got what you mean, but everything which is stored is not created due to batches. So job created for example is not written in a batch and also the workflow instance creation is not.

Reply:
Ah true, we only store for processes and jobs, so they are never part of a batch (I think? Can't remember how a multi-instance call activity would look like :sweat_smile:)

Reply:
Even if multi instances creates multiple jobs the CREATED will be written again one by one :slightly_smiling_face: so should be fine